### PR TITLE
Adds Apos pages to the Gatsby GraphQL API

### PR DESCRIPTION
**Pull request by abea**
Pages are added, including (by default) a `_rendered` property with the rendered refreshable HTML.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,6 +32,38 @@ module.exports = {
...
+
+An ApostropheCMS API key with permissions to make GET requests. [Read more about setting up API keys](https://a3.docs.apos.dev/reference/api/authentication.html#api-keys) in the Apostrophe documentation.
+
+### baseUrl (required)
+
+The base URL for API requests. Usually this will be the root domain of the Apostrophe website or application with no trailing slash. In local development it will be `http://localhost:3000` by default.
+
+### pieceTypes
+
+An array of Apostrophe piece type names. These will the same as the key name entered in the `app.js` `modules` object in the Apostrophe app. You may include core piece types that are not included by default, e.g, `apostrophe/file`, `apostrophe/image-tag`.
```
Comment (boutell) [inline README.md]: That's `@apostrophecms/file` and `@apostrophecms/image-tag`.

File: lib/pieces.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+async function generatePieceNodes (pieces, nodeType, utils) {
+  // loop through data and create Gatsby nodes
+  pieces.forEach(piece => {
+    generateNode(piece, nodeType, utils);
+  });
+}
+
+function isAposCoreType(name) {
+  return name.length > 15 && name.slice(0, 15) === '@apostrophecms/';
```
Comment (boutell) [inline lib/pieces.js]: What happens to modules from other npm namespaces?

File: lib/pages.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      }
+
+      if (!page || page.message === 'notfound') {
+        // This is most likely the trash "page"
+        console.warn(`No page found for Apostrophe document ID ${id}.`);
+        return;
+      }
+
+      if (renderPages !== false && page._url) {
+        await addRenderedContent(page, baseUrl);
```
Comment (boutell) [inline lib/pages.js:37]: You might want this for a piece that has a _url, too. I agree it is somewhat more probable in the case of a page.

**Review by boutell — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -32,6 +32,38 @@ module.exports = {
...
+
+### Rendered content
+
+Apostrophe pieces are well suited to being delivered as structured, JSON-like data. Apostrophe pages, on the other hand, usually consist primarily of the flexible content "areas," populated with any number of different widget types. It powers Apostrophe's great in-context editing experience, but the raw data is not easy to use in a page template on your own.
+
+Unless you add `renderPages: false` to this source plugin's options, pages will appear with a `_rendered` property in Gatsby GraphQL queries. This property's value is a string of HTML, rendered using the relevant page template in the Apostrophe app. That HTML can be used in your Gatsby site to [programmatically create pages](https://www.gatsbyjs.com/docs/tutorial/part-seven/) using the right layout component and slug structure for your site.
+
 ## Planned features
 
 - Apos "areas" are rendered as HTML by default, rather than output as less usable raw data
```
Comment (boutell) [inline README.md:69]: ... Oh I see, this is still to come (separate PR?)

File: lib/pages.js (hunk 3 above)
Comment (abea) [reply lib/pages.js:37]: I think we should hold off on that for now. For one, I'm thinking of the Gatsby use case. It seems to be all about that GraphQL. Simply delivering rendered HTML doesn't seem to be The Way. It also adds a ton of more requests. I'd rather this be happening on the Apostrophe end if we want to do more rendering.

File: README.md (hunk 4 above)
Comment (abea) [reply README.md:69]: Yep, it's a PR in core now. I'll then make a PR here, adding the query parameter.

File: lib/pieces.js (hunk 2 above)
Comment (abea) [reply lib/pieces.js]: Is a good question. Those we could simply strip away the punctuation.

**Review by boutell — APPROVED** [review]

**Pull request merged**